src/index.coffee

```diff
@@ -0,0 +1,90 @@
+EvalSha = require 'redis-evalsha'
+_ = require 'underscore'
+async = require 'async'
+fs = require 'fs'
+redis = require 'redis'
+
+module.exports = class RateLimit
+
+  constructor: (@redisClient, rules, @prefix = 'ratelimit') ->
+    @checkFn = 'check_rate_limit'
+    @checkIncrFn = 'check_incr_rate_limit'
+
+    @eval = new EvalSha @redisClient
+    @eval.add @checkFn, @checkLimitScript()
+    @eval.add @checkIncrFn, @checkLimitIncrScript()
+
+    @rules = @convertRules rules
+
+  readLua: (filename) ->
+    fs.readFileSync("#{__dirname}/../lua/#{filename}.lua").toString()
+
+  checkLimitScript: ->
+    [
+      @readLua 'unpack_args'
+      @readLua 'check_limit'
+      'return 0'
+    ].join '\n'
+
+  checkLimitIncrScript: ->
+    [
+      @readLua 'unpack_args'
+      @readLua 'check_limit'
+      @readLua 'check_incr_limit'
+    ].join '\n'
+
+  convertRules: (rules) ->
+    for rule in rules
+      [rule.interval, rule.limit]
+
+  scriptArgs: (keys, weight = 1) ->
+    # Keys has to be a list
+    adjustedKeys = if _.isArray(keys) then keys else [keys]
+    adjustedKeys = _.chain(adjustedKeys)
+      .compact()
+      .filter(_.isString)
+      .filter (key) ->
+        key.length > 0
+      .map (key) =>
+        "#{@prefix}:#{key}"
+      .value()
+
+    throw new Error "Bad keys: #{keys}" if adjustedKeys.length is 0
+
+    rules = JSON.stringify @rules
+    ts = Math.floor(Date.now() / 1000)
+    [adjustedKeys, [rules, ts, weight]]
+
+  check: (keys, callback) ->
+    try
+      [keys, args] = @scriptArgs keys
+      @eval.exec @checkFn, keys, args, (err, result) ->
```

Comment:
Nesting an asynchronous function in a `try .. catch` block is probably not a good idea - I would write this as the following:

```coffee
check: (keys, callback) ->
  try
    [keys, args] = @scriptArgs keys
  catch err
    return callback err

  @eval.exec @checkFn, keys, args, (err, result) ->
    callback err, result is 1
```